packages/hasher/src/repoInfo.ts

```diff
@@ -0,0 +1,74 @@
+import {
+  WorkspaceInfo,
+  ParsedLock,
+  findGitRoot,
+  getWorkspaces,
+  parseLockFile
+} from "workspace-tools";
+
+import { getPackageDeps } from "@rushstack/package-deps-hash";
+
+export interface RepoInfo {
+  root: string;
+  workspaceInfo: WorkspaceInfo;
+  parsedLock: ParsedLock;
+  repoHashes: { [key: string]: string };
+}
+
+const repoInfoCache: RepoInfo[] = [];
+
+/**
+ * repoInfo cache lookup - it is specialized to be using a substring match to make it run as fast as possible
+ * @param packageRoot
+ */
+function searchRepoInfoCache(packageRoot: string) {
+  for (const repoInfo of repoInfoCache) {
+    if (repoInfo.workspaceInfo && packageRoot.startsWith(repoInfo.root)) {
+      return repoInfo;
+    }
+  }
+}
+
+export async function _getRepoInfoNoCache(cwd: string) {
+  const root = findGitRoot(cwd);
+  if (!root) {
+    throw new Error("Cannot initialize Repo class without a .git root");
+  }
+
+  const repoHashes = getPackageDeps(root).files;
+  const workspaceInfo = await getWorkspaces(root);
+  const parsedLock = await parseLockFile(root);
+
+  const repoInfo = {
+    root,
+    workspaceInfo,
+    parsedLock,
+    repoHashes
+  };
+
+  repoInfoCache.push(repoInfo);
+  return repoInfo;
+}
+
+// A promise to guarantee the getRepoInfo is done one at a time
```

Comment:
Basically, the oneAtATime promise turn this asynchronous getRepoInfo() function to be a **non-blocking** function to perform repoInfo cache hit checking one at a time. In a p-queue situation, this function gets queued up and is bombarded to be called. If I don't have some mechanism to make sure to call this one at a time, we'll have a race condition of N calls to this and creating the cache. So, it'll end up calling the getPackageDeps() many times before being able to take advantage of a in-memory cache of it. It is super fast because cache hit is 100% basically given that all packageRoots are in the workspace.

I THINK I had explained this as much as I can in the getRepoInfo comments below, but if you can think of a way to explain it better please add a suggestion here.

Reply:
I still don't understand, I am confused by the word parallel, I thought that nothing would run in parallel in NodeJS unless you use workers. Or is the issue that the function is recursive? I am confused because I have never seen this pattern, it looks bizarre to me so I suspect there is a simpler way to achieve the same thing.
This is just my feedback, I won't block the PR if you choose to dismiss it :)